File: Codes/2512-reward-top-k-students.cpp
/*
 * @Author: LetMeFly
 * @Date: 2023-10-11 09:40:06
 * @LastEditors: LetMeFly
 * @LastEditTime: 2023-10-11 09:50:17
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    vector<int> topStudents(vector<string>& positive_feedback, vector<string>& negative_feedback, vector<string>& report, vector<int>& student_id, int k) {
        unordered_set<string> postive, negative;
        for (string& s : positive_feedback) {
            postive.insert(s);
        }
        for (string& s : negative_feedback) {
            negative.insert(s);
        }
        vector<pair<int, int>> v;
        for (int i = 0; i < report.size(); i++) {
            int lastSpace = -1;
            int thisScore = 0;
            for (int j = 0; j <= report[i].size(); j++) {
                if (j == report[i].size() || report[i][j] == ' ') {  // (lastSpace, j)
                    string thisWord = report[i].substr(lastSpace + 1, j - lastSpace - 1);
                    lastSpace = j;
                    if (postive.count(thisWord)) {
                        thisScore += 3;
                    }
                    else if (negative.count(thisWord)) {
                        thisScore -= 1;
                    }
                }
            }
            v.push_back({-thisScore, student_id[i]});
        }
        sort(v.begin(), v.end());
        vector<int> ans;
        for (int i = 0; i < k; i++) {
            ans.push_back(v[i].second);
        }
        return ans;
    }
};

File: Codes/2512-reward-top-k-students.py
'''
Author: LetMeFly
Date: 2023-10-11 09:52:04
LastEditors: LetMeFly
LastEditTime: 2023-10-11 09:54:44
'''
from typing import List

class Solution:
    def topStudents(self, positive_feedback: List[str], negative_feedback: List[str], report: List[str], student_id: List[int], k: int) -> List[int]:
        mp = {}
        for s in positive_feedback:
            mp[s] = -3
        for s in negative_feedback:
            mp[s] = 1
        v = []
        for i in range(len(report)):
            v.append([sum(mp.get(word, 0) for word in report[i].split()), student_id[i]])
        v.sort()
        return [t[1] for t in v[:k]]